Getting query error on Postgres w/ timezone set

Looks like this just started after the recent changes with the fix for timezone handling on MySQL.

```
2015-11-02 15:26:13,321 [ERROR] metabase.driver.generic-sql.query-processor :: Failed to set timezone: Batch entry 0 SET LOCAL timezone TO 'US/Pacific' was aborted.  Call getNextException to see the cause.
2015-11-02 15:26:13,322 [WARN ] com.mchange.v2.c3p0.impl.NewPooledConnection :: [c3p0] A PooledConnection that has already signalled a Connection error is still in use!
2015-11-02 15:26:13,322 [WARN ] com.mchange.v2.c3p0.impl.NewPooledConnection :: [c3p0] Another error has occurred [ org.postgresql.util.PSQLException: ERROR: current transaction is aborted, commands ignored until end of transaction block ] which will not be reported to listeners!
org.postgresql.util.PSQLException: ERROR: current transaction is aborted, commands ignored until end of transaction block
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2198)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1927)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:255)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:561)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:419)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeQuery(AbstractJdbc2Statement.java:304)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:76)
	at clojure.java.jdbc$db_query_with_resultset$run_query_with_params__6847.invoke(jdbc.clj:824)
	at clojure.java.jdbc$db_query_with_resultset.invoke(jdbc.clj:831)
	at clojure.java.jdbc$query.doInvoke(jdbc.clj:866)
	at clojure.lang.RestFn.invoke(RestFn.java:464)
	at korma.db$exec_sql.invoke(db.clj:259)
	at korma.db$do_query.invoke(db.clj:275)
	at korma.core$exec.invoke(core.clj:498)
	at metabase.driver.generic_sql.query_processor$process_structured$fn__27494.invoke(query_processor.clj:242)
	at clojure.java.jdbc$db_transaction_STAR_.doInvoke(jdbc.clj:595)
	at clojure.lang.RestFn.invoke(RestFn.java:521)
	at metabase.driver.generic_sql.query_processor$process_structured.invoke(query_processor.clj:238)
	at metabase.driver.generic_sql.query_processor$process_and_run.invoke(query_processor.clj:257)
	at metabase.driver.generic_sql$process_query.invoke(generic_sql.clj:28)
	at metabase.driver.query_processor$wrap_guard_multiple_calls$fn__25530.invoke(query_processor.clj:240)
	at metabase.driver.query_processor$pre_log_query$fn__25524.invoke(query_processor.clj:199)
	at metabase.driver.query_processor.annotate$post_annotate$fn__24872.invoke(annotate.clj:271)
	at metabase.driver.query_processor$limit$fn__25514.invoke(query_processor.clj:181)
	at metabase.driver.query_processor$cumulative_sum$fn__25508.invoke(query_processor.clj:168)
	at metabase.driver.query_processor$pre_add_implicit_breakout_order_by$fn__25465.invoke(query_processor.clj:104)
	at metabase.driver.query_processor$pre_add_implicit_fields$fn__25446.invoke(query_processor.clj:82)
	at metabase.driver.query_processor$post_add_row_count_and_status$fn__25432.invoke(query_processor.clj:64)
	at metabase.driver.query_processor$pre_expand$fn__25429.invoke(query_processor.clj:57)
	at metabase.driver.query_processor$wrap_catch_exceptions$fn__25422.invoke(query_processor.clj:43)
	at metabase.driver.query_processor$process_structured.invoke(query_processor.clj:246)
	at metabase.driver.query_processor$process.invoke(query_processor.clj:275)
	at metabase.driver$process_query.invoke(driver.clj:385)
	at metabase.driver$dataset_query.invoke(driver.clj:425)
	at metabase.api.dataset$fn__26018$fn__26021.invoke(dataset.clj:14)
	at metabase.api.common.internal$wrap_catch_api_exceptions.invoke(internal.clj:216)
	at metabase.api.dataset$fn__26018.invoke(dataset.clj:10)
	at compojure.core$make_route$fn__3086.invoke(core.clj:130)
	at compojure.core$wrap_route_middleware$fn__3082.invoke(core.clj:118)
	at compojure.core$if_route$fn__3054.invoke(core.clj:41)
	at compojure.core$if_method$fn__3046.invoke(core.clj:27)
	at compojure.core$routing$fn__3093.invoke(core.clj:144)
	at clojure.core$some.invoke(core.clj:2570)
	at compojure.core$routing.doInvoke(core.clj:144)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invoke(core.clj:632)
	at compojure.core$routes$fn__3097.invoke(core.clj:149)
	at metabase.middleware$enforce_authentication$fn__17271.invoke(middleware.clj:73)
	at compojure.core$routing$fn__3093.invoke(core.clj:144)
	at clojure.core$some.invoke(core.clj:2570)
	at compojure.core$routing.doInvoke(core.clj:144)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at metabase.api.routes$fn__26763.invoke(routes.clj:35)
	at compojure.core$if_context$fn__3155.invoke(core.clj:211)
	at compojure.core$routing$fn__3093.invoke(core.clj:144)
	at clojure.core$some.invoke(core.clj:2570)
	at compojure.core$routing.doInvoke(core.clj:144)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invoke(core.clj:632)
	at compojure.core$routes$fn__3097.invoke(core.clj:149)
	at compojure.core$routing$fn__3093.invoke(core.clj:144)
	at clojure.core$some.invoke(core.clj:2570)
	at compojure.core$routing.doInvoke(core.clj:144)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at metabase.routes$fn__26802.invoke(routes.clj:29)
	at compojure.core$if_context$fn__3155.invoke(core.clj:211)
	at compojure.core$routing$fn__3093.invoke(core.clj:144)
	at clojure.core$some.invoke(core.clj:2570)
	at compojure.core$routing.doInvoke(core.clj:144)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invoke(core.clj:632)
	at compojure.core$routes$fn__3097.invoke(core.clj:149)
	at metabase.middleware$log_api_call$fn__17363.invoke(middleware.clj:235)
	at metabase.middleware$add_security_headers$fn__17300.invoke(middleware.clj:123)
	at metabase.middleware$format_response$fn__17328.invoke(middleware.clj:178)
	at ring.middleware.json$wrap_json_body$fn__6195.invoke(json.clj:44)
	at ring.middleware.json$wrap_json_response$fn__6212.invoke(json.clj:87)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__6246.invoke(keyword_params.clj:32)
	at ring.middleware.params$wrap_params$fn__3264.invoke(params.clj:57)
	at metabase.middleware$bind_current_user$fn__17275.invoke(middleware.clj:87)
	at metabase.middleware$wrap_current_user_id$fn__17249.invoke(middleware.clj:55)
	at metabase.middleware$wrap_api_key$fn__17290.invoke(middleware.clj:98)
	at metabase.middleware$wrap_session_id$fn__17242.invoke(middleware.clj:46)
	at ring.middleware.cookies$wrap_cookies$fn__5304.invoke(cookies.clj:171)
	at ring.middleware.session$wrap_session$fn__6371.invoke(session.clj:85)
	at ring.middleware.gzip$wrap_gzip$fn__5341.invoke(gzip.clj:65)
	at clojure.lang.Var.invoke(Var.java:379)
	at ring.middleware.reload$wrap_reload$fn__1419.invoke(reload.clj:22)
	at ring.middleware.stacktrace$wrap_stacktrace_log$fn__854.invoke(stacktrace.clj:23)
	at ring.middleware.stacktrace$wrap_stacktrace_web$fn__916.invoke(stacktrace.clj:86)
	at ring.adapter.jetty$proxy_handler$fn__99.invoke(jetty.clj:24)
	at ring.adapter.jetty.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:497)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
2015-11-02 15:26:13,325 [DEBUG] metabase.db.internal :: DB CALL:  Database [:name :id :engine :details] (where (limit (select* ENTITY25341) 1) {:id database-id})
2015-11-02 15:26:13,326 [DEBUG] metabase.db.internal :: DB CALL:  Table [:name :id] (where (select* ENTITY25403) {:id [in table-ids]})
2015-11-02 15:26:13,328 [ERROR] metabase.driver :: Query failure: ERROR: current transaction is aborted, commands ignored until end of transaction block
```

--- a/src/metabase/driver/generic_sql/native.clj
+++ b/src/metabase/driver/generic_sql/native.clj
@@ -25,26 +25,30 @@
              {:keys [features set-timezone-sql]} (driver/engine->driver (:engine database))]
 
          (jdbc/with-db-transaction [t-conn db-conn]
+           (let [^java.sql.Connection jdbc-connection (:connection t-conn)]
+             ;; Disable auto-commit for this transaction, that way shady queries are unable to modify the database
+             (.setAutoCommit jdbc-connection false)
+             (try
+               ;; Set the timezone if applicable
+               (when-let [timezone (driver/report-timezone)]
+                 (when (and (seq timezone)
+                            (contains? features :set-timezone))
+                   (log/debug (u/format-color 'green "%s" set-timezone-sql))
+                   (try (jdbc/db-do-prepared t-conn set-timezone-sql [timezone])
+                        (catch Throwable e
+                          (log/error (u/format-color 'red "Failed to set timezone: %s" (.getMessage e)))))))
 
-           ;; Set the timezone if applicable. We do this *before* making the transaction read-only because some DBs
-           ;; won't let you set the timezone on a read-only connection
-           (when-let [timezone (driver/report-timezone)]
-             (when (and (seq timezone)
-                        (contains? features :set-timezone))
-               (log/debug (u/format-color 'green "%s" set-timezone-sql))
-               (try (jdbc/db-do-prepared t-conn set-timezone-sql [timezone])
-                    (catch Throwable e
-                      (log/error (u/format-color 'red "Failed to set timezone: %s" (.getMessage e)))))))
+               ;; Now run the query itself
+               (log/debug (u/format-color 'green "%s" sql))
+               (let [[columns & [first-row :as rows]] (jdbc/query t-conn sql, :as-arrays? true)]
+                 {:rows    rows
+                  :columns columns
+                  :cols    (for [[column first-value] (zipmap columns first-row)]
+                             {:name      column
+                              :base_type (value->base-type first-value)})})
 
-           ;; Now make the transaction read-only and run the query itself
-           (.setReadOnly ^com.mchange.v2.c3p0.impl.NewProxyConnection (:connection t-conn) true)
-           (log/debug (u/format-color 'green "%s" sql))
-           (let [[columns & [first-row :as rows]] (jdbc/query t-conn sql, :as-arrays? true)]
-             {:rows    rows
-              :columns columns
-              :cols    (for [[column first-value] (zipmap columns first-row)]
-                         {:name      column
-                          :base_type (value->base-type first-value)})})))
+               ;; Rollback any changes made during this transaction just to be extra-double-sure JDBC doesn't try to commit them automatically for us
+               (finally (.rollback jdbc-connection))))))
        (catch java.sql.SQLException e
          (let [^String message (or (->> (.getMessage e) ; error message comes back like 'Column "ZID" not found; SQL statement: ... [error-code]' sometimes
                                         (re-find #"^(.*);") ; the user already knows the SQL, and error code is meaningless

--- a/src/metabase/driver/generic_sql/query_processor.clj
+++ b/src/metabase/driver/generic_sql/query_processor.clj
@@ -1,17 +1,18 @@
 (ns metabase.driver.generic-sql.query-processor
   "The Query Processor is responsible for translating the Metabase Query Language into korma SQL forms."
   (:require [clojure.core.match :refer [match]]
-            [clojure.tools.logging :as log]
+            [clojure.java.jdbc :as jdbc]
             [clojure.string :as s]
+            [clojure.tools.logging :as log]
             (korma [core :as k]
                    [db :as kdb])
             (korma.sql [fns :as kfns]
                        [utils :as utils])
             [metabase.config :as config]
             [metabase.driver :as driver]
-            [metabase.driver.query-processor :as qp]
             (metabase.driver.generic-sql [native :as native]
                                          [util :refer :all])
+            [metabase.driver.query-processor :as qp]
             [metabase.util :as u])
   (:import java.sql.Timestamp
            java.util.Date
@@ -235,17 +236,18 @@
         (kdb/with-db (:db entity)
           (if (and (seq timezone)
                    (contains? (:features driver) :set-timezone))
-            (kdb/transaction
-             (try (k/exec-raw [(:set-timezone-sql driver) [timezone]])
-                  (catch Throwable e
-                    (log/error (u/format-color 'red "Failed to set timezone: %s" (.getMessage e)))))
-             (k/exec korma-query))
+            (try (kdb/transaction (k/exec-raw [(:set-timezone-sql driver) [timezone]])
+                                  (k/exec korma-query))
+                 (catch Throwable e
+                   (log/error (u/format-color 'red "Failed to set timezone:\n%s"
+                                (with-out-str (jdbc/print-sql-exception-chain e))))
+                   (k/exec korma-query)))
             (k/exec korma-query))))
 
       (catch java.sql.SQLException e
-        (let [^String message (or (->> (.getMessage e) ; error message comes back like "Error message ... [status-code]" sometimes
+        (let [^String message (or (->> (.getMessage e)                       ; error message comes back like "Error message ... [status-code]" sometimes
                                        (re-find  #"(?s)(^.*)\s+\[[\d-]+\]$") ; status code isn't useful and makes unit tests hard to write so strip it off
-                                       second) ; (?s) = Pattern.DOTALL - tell regex `.` to match newline characters as well
+                                       second)                               ; (?s) = Pattern.DOTALL - tell regex `.` to match newline characters as well
                                   (.getMessage e))]
           (throw (Exception. message)))))))
 

--- a/src/metabase/driver/postgres.clj
+++ b/src/metabase/driver/postgres.clj
@@ -196,7 +196,7 @@
        :unix-timestamp->timestamp         unix-timestamp->timestamp
        :date                              date
        :date-interval                     date-interval
-       :set-timezone-sql                  "SET LOCAL timezone TO ?;"
+       :set-timezone-sql                  "UPDATE pg_settings SET setting = ? WHERE name ILIKE 'timezone';"
        :driver-specific-sync-field!       driver-specific-sync-field!
        :humanize-connection-error-message humanize-connection-error-message}
       sql-driver

